plugins/rest/aws_test.go

```diff
@@ -1092,6 +1093,60 @@ func TestWebIdentityCredentialService(t *testing.T) {
 	})
 }
 
+func TestCustomIdentityCredentialService(t *testing.T) {
+	t.Setenv("AWS_REGION", "us-west-1")
+
+	testAccessKey := "ASgeIAIOSFODNN7EXAMPLE"
+	ts := stsTestServer{
+		t:                         t,
+		accessKey:                 testAccessKey,
+		assumeRoleWithWebIdentity: false,
+	}
+	ts.start()
+	defer ts.stop()
+	cs := awsCustomIdentityCredentialService{
+		stsURL: ts.server.URL,
+		logger: logging.Get(),
+	}
+
+	// wrong path: no AWS_ROLE_ARN set
+	err := cs.populateFromEnv()
+	assertErr("no AWS_ROLE_ARN set in environment", err, t)
+	t.Setenv("AWS_ROLE_ARN", "role:arn")
+
+	// wrong path: no AWS_ACCESS_KEY_ID set
+	err = cs.populateFromEnv()
+	assertErr("no AWS_ACCESS_KEY_ID set in environment", err, t)
+	t.Setenv("AWS_ACCESS_KEY_ID", "MYAWSACCESSKEYGOESHERE")
+
+	// wrong path: no AWS_SECRET_ACCESS_KEY set
+	err = cs.populateFromEnv()
+	assertErr("no AWS_SECRET_ACCESS_KEY set in environment", err, t)
+	t.Setenv("AWS_SECRET_ACCESS_KEY", "MYAWSSECRETACCESSKEYGOESHERE")
+
+	// happy path: env vars set
+	err = cs.populateFromEnv()
+	if err != nil {
+		t.Fatalf("Error while getting env vars: %s", err)
+	}
+
+	// happy path: refresh and get credentials
+	creds, _ := cs.credentials(context.Background())
+	assertEq(creds.AccessKey, testAccessKey, t)
+
+	// happy path: refresh with session and get credentials
+	cs.expiration = time.Now()
+	cs.SessionName = "TEST_SESSION"
+	creds, _ = cs.credentials(context.Background())
+	assertEq(creds.AccessKey, testAccessKey, t)
+
+	// happy path: don't refresh as credentials not expired so STS not called
+	// verify existing credentials haven't changed
+	ts.accessKey = "OTHERKEY"
+	creds, _ = cs.credentials(context.Background())
+	assertEq(creds.AccessKey, testAccessKey, t)
```

Comment:
Consider adding another pair of assertions that we get the expected refresh/no refresh just after and before the hard-coded 5 minutes grace time.

Reply:
Good idea. Added a test case.